### PR TITLE
feat: customized canvas entry

### PR DESCRIPTION
**Pull request by yy-wow**
English | [简体中文](https://github.com/opentiny/tiny-engine/blob/develop/.github/PULL_REQUEST_TEMPLATE/PULL_REQUEST_TEMPLATE.zh-CN.md)

# PR

## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit message follows our [Commit Message Guidelines](https://github.com/opentiny/tiny-engine/blob/develop/CONTRIBUTING.md)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)
- [ ] Built its own designer, fully self-validated

## PR Type

What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

### What is the current behavior?

Issue Number: N/A

### What is the new behavior?
1. 工具栏-保存增加beforeSave、saveMethod、saved钩子；
2. 工具栏-预览增加beforePreview、previewMethod、afterPreview钩子；
3. 画布iframe支持配置src替换默认的srcdoc，支持自定义的画布页面构建入口；
4. 导出获取组件condition值的方法；
5. 修复一些eslint报错问题

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

## Release Notes

- **New Features**
  - Enhanced iframe rendering logic in the Canvas components to support conditional sources.
  - Introduced new props for better handling of canvas sources and visibility states in various components.
  - Added new methods for managing conditions and improved state management capabilities.
  - Added a new `createRender` function to facilitate rendering operations.
  - Expanded library entry points to include a new `render` entry.

- **Improvements**
  - Asynchronous handling added to several functions, improving the flow of operations during save and preview actions.
  - Added pre-save and post-save hooks for greater flexibility in handling method execution.
  - Streamlined control flow for save operations, enhancing user experience during loading states.
  - Improved interactivity and maintainability of components through method restructuring.

- **Bug Fixes**
  - Resolved issues related to method execution and state management within the components.

- **Configuration Changes**
  - Updated project to use ECMAScript modules, enhancing module import and export syntax.
  - Added `"type": "module"` property in `package.json` for modern JavaScript standards.
  - Enhanced module export structure with new entries in `package.json`.

### CODE REVIEW

File: packages/canvas/index.js [hunk 1]
```diff
@@ -17,6 +17,7 @@ import CanvasLayout from './layout'
 import DesignCanvas from './DesignCanvas'
 
 export { CanvasContainer, CanvasLayout, DesignCanvas }
+export { createRender } from './render'
```
[the remark this answers is not in the document]
Comment (yy-wow) [reply packages/canvas/index.js]: from './render/index'

File: packages/toolbars/preview/src/Main.vue [hunk 2]
```diff
@@ -13,7 +13,8 @@
 <script>
 import { previewPage, previewBlock } from '@opentiny/tiny-engine-common/js/preview'
 import { useBlock, useCanvas, useLayout, useNotify } from '@opentiny/tiny-engine-meta-register'
-import { getMergeMeta } from '@opentiny/tiny-engine-meta-register'
+import { getMergeMeta, getOptions } from '@opentiny/tiny-engine-meta-register'
```
[the remark this answers is not in the document]
Comment (yy-wow) [reply packages/toolbars/preview/src/Main.vue:16]: used at line 66

File: packages/toolbars/preview/src/Main.vue [hunk 3]
```diff
@@ -30,7 +31,26 @@
     const { isBlock, getCurrentPage, canvasApi } = useCanvas()
     const { getCurrentBlock } = useBlock()
 
-    const preview = () => {
+    const preview = async () => {
+      const { beforePreview, previewMethod, afterPreview } = getOptions(meta.id)
+      const printer = console
```
[the remark this answers is not in the document]
Comment (yy-wow) [reply packages/toolbars/preview/src/Main.vue]: 修改为useNotify输出错误信息

File: package.json [hunk 4]
```diff
@@ -1,6 +1,7 @@
 {
   "name": "tiny-engine",
   "private": true,
+  "type": "module",
```
[the remark this answers is not in the document]
Comment (hexqi) [reply package.json]: 外层应该不用改成type:module，如果改外层的话根目录和scripts里面所有脚本都需要改后缀。 看报错是canvas包里面eslintrc文件需要改成cjs后缀

**Review by chilingling — DISMISSED** [review]

File: package.json (hunk 4 above)
[the remark this answers is not in the document]
Comment (yy-wow) [reply package.json]: 已修改

**Review by chilingling — DISMISSED** [review]

**Review by hexqi — DISMISSED** [review]

File: packages/canvas/vite.config.js [hunk 5]
```diff
@@ -34,7 +34,8 @@ export default defineConfig({
     cssCodeSplit: true,
     lib: {
       entry: {
-        index: path.resolve(__dirname, './index.js')
+        index: path.resolve(__dirname, './index.js'),
+        render: path.resolve(__dirname, './render/index.js'),
       },
       name: 'canvas',
       fileName: () => 'index.js',
```
Comment (chilingling) [inline packages/canvas/vite.config.js]: 这里  `fileName` 也要同步改

Comment (yy-wow) [reply packages/canvas/vite.config.js]: rollupOptions.output比lib.filename优先级高，被覆盖，此处可删除

**Review by rhlin — APPROVED** [review]
LGTM

**Review by hexqi — APPROVED** [review]

**Pull request merged**